pennylane_lightning/core/src/simulators/lightning_kokkos/catalyst/LightningKokkosSimulator.cpp

```diff
@@ -0,0 +1,575 @@
+            *(samplesIter++) =
+                static_cast<double>(li_samples[shot * numQubits + wire]);
+        }
+    }
+}
+
+void LightningKokkosSimulator::Counts(DataView<double, 1> &eigvals,
+                                      DataView<int64_t, 1> &counts,
+                                      size_t shots) {
+    const size_t numQubits = this->GetNumQubits();
+    const size_t numElements = 1U << numQubits;
+
+    RT_FAIL_IF(eigvals.size() != numElements || counts.size() != numElements,
+               "Invalid size for the pre-allocated counts");
+
+    // generate_samples is a member function of the MeasuresKokkos class.
+    Pennylane::LightningKokkos::Measures::Measurements<StateVectorT> m{
+        *(this->device_sv)};
+
+    // PL-Lightning-Kokkos generates samples using the alias method.
+    // Reference: https://en.wikipedia.org/wiki/Inverse_transform_sampling
+    auto li_samples = m.generate_samples(shots);
+
+    // Fill the eigenvalues with the integer representation of the corresponding
+    // computational basis bitstring. In the future, eigenvalues can also be
+    // obtained from an observable, hence the bitstring integer is stored as a
+    // double.
+    std::iota(eigvals.begin(), eigvals.end(), 0);
+    std::fill(counts.begin(), counts.end(), 0);
+
+    // The lightning samples are layed out as a single vector of size
+    // shots*qubits, where each element represents a single bit. The
+    // corresponding shape is (shots, qubits). Gather the bits of all qubits
+    // into a bitstring.
+    for (size_t shot = 0; shot < shots; shot++) {
+        std::bitset<CHAR_BIT * sizeof(double)> basisState;
+        size_t idx = numQubits;
+        for (size_t wire = 0; wire < numQubits; wire++) {
+            basisState[--idx] = li_samples[shot * numQubits + wire];
+        }
+        counts(static_cast<size_t>(basisState.to_ulong())) += 1;
+    }
+}
+
+void LightningKokkosSimulator::PartialCounts(
+    DataView<double, 1> &eigvals, DataView<int64_t, 1> &counts,
+    const std::vector<QubitIdType> &wires, size_t shots) {
+    const size_t numWires = wires.size();
+    const size_t numQubits = this->GetNumQubits();
+    const size_t numElements = 1U << numWires;
+
+    RT_FAIL_IF(numWires > numQubits, "Invalid number of wires");
+    RT_FAIL_IF(!isValidQubits(wires), "Invalid given wires to measure");
+    RT_FAIL_IF((eigvals.size() != numElements || counts.size() != numElements),
+               "Invalid size for the pre-allocated partial-counts");
+
+    // get device wires
+    auto &&dev_wires = getDeviceWires(wires);
+
+    // generate_samples is a member function of the MeasuresKokkos class.
+    Pennylane::LightningKokkos::Measures::Measurements<StateVectorT> m{
+        *(this->device_sv)};
+
+    // PL-Lightning-Kokkos generates samples using the alias method.
+    // Reference: https://en.wikipedia.org/wiki/Inverse_transform_sampling
+    auto li_samples = m.generate_samples(shots);
```

Comment:
Wish we could pass a `wires` argument in here.

Reply:
yes!